### PR TITLE
Throw BadImageFormatException when missing PE metadata

### LINKED ISSUE #6200: ResolveAssemblyReference throws InvalidOperationException on invalid PE files
### Issue Description
I have a project which produces native binaries (on linux, so ELF format) and when I use ProjectReference to indicate project dependencies, ResolveAssemblyReference tries to read the metadata of these files and throws InvalidOperationException because the dependencies don't use the PE format, or have metadata.

### Steps to Reproduce
Reproducing this is pretty hard in my experience, since *most* ELF files make the PEReader throw a BadImageFormatException. I've opened a related issue in the .NET runtime because of this: https://github.com/dotnet/runtime/issues/48419

The summary is, though:
1. Create a MSBuild project which produces a non-PE format file (e.g. by writing a large text file to `$(IntermediateOutputPath)$(TargetName)$(TargetExt)`)
2. Try to reference that project from another using `ProjectReference`
3. Build the project in step 2

### Expected Behavior
ResolveAssemblyReference skips the project because it doesn't have any metadata

### Actual Behavior
ResolveAssemblyReference throws an InvalidOperationException from PEReader.GetMetadataBlock()

### Analysis
Doing some digging, this appears to be caused by [AssemblyInformation.CorePopulateMetadata](https://github.com/dotnet/msbuild/blob/ca44138662e3aa90eb9305dd31d906ef02e962cb/src/Tasks/AssemblyDependency/AssemblyInformation.cs#L362) not checking whether the created PEReader has any metadata. This appears to be checked everywhere else a PEReader is used:
- https://github.com/dotnet/msbuild/blob/ca44138662e3aa90eb9305dd31d906ef02e962cb/src/Tasks/ManifestUtil/MetadataReader.cs#L35-L42
- https://github.com/dotnet/msbuild/blob/6819f7ab06c3f43e83ff4059d417395e0af06c01/src/Shared/AssemblyNameExtension.cs#L214-L220

### Versions & Configurations
```
$ dotnet --info
.NET SDK (reflecting any global.json):
 Version:   5.0.103
 Commit:    9effbc8ad5

Runtime Environment:
 OS Name:     debian
 OS Version:  
 OS Platform: Linux
 RID:         debian-x64
 Base Path:   /usr/share/dotnet/sdk/5.0.103/

Host (useful for support):
  Version: 5.0.3
  Commit:  eae88cc11b

.NET SDKs installed:
  3.1.406 [/usr/share/dotnet/sdk]
  5.0.103 [/usr/share/dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 3.1.12 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.3 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.1.12 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.3 [/usr/share/dotnet/shared/Microsoft.NETCore.App]

To install additional .NET runtimes or SDKs:
  https://aka.ms/dotnet-download
$ dotnet msbuild -version
Microsoft (R) Build Engine version 16.8.3+39993bd9d for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

16.8.3.61104
```

### Attach a binlog
[This binlog](https://github.com/dotnet/msbuild/files/6052440/msbuild.binlog.zip) reproduces the issue, and I've been unable to reproduce it using smaller projects. I think this may be because smaller projects produce smaller binaries, which PEReader ends up rejecting. The source for the `ClangLink` and `ClangCompile` tasks can be found [here](https://github.com/finitereality/finite.cpp.sdk).

### PR DESCRIPTION
**Pull request by FiniteReality**
Fixes #6200 

### Context
`InvalidOperationException` was being thrown by `AssemblyInformation.CorePopulateMetadata`, causing the ResolveAssemblyReference task to fail rather than disregarding the file.

### Changes Made
I copied the code from AssemblyNameExtension to ensure there is metadata present before getting a metadata reader: https://github.com/dotnet/msbuild/blob/6819f7ab06c3f43e83ff4059d417395e0af06c01/src/Shared/AssemblyNameExtension.cs#L214-L231

### Testing
I looked for unit tests for AssemblyInformation, but could not find any, so I didn't add a unit test for this case. However, I successfully built the project that was reproducing the error in #6200, indicating that the change did indeed fix the issue.

### CODE REVIEW

File: src/Tasks/AssemblyDependency/AssemblyInformation.cs [hunk 1]
```diff
@@ -361,6 +361,21 @@ private void CorePopulateMetadata()
                 using (var stream = File.OpenRead(_sourceFile))
                 using (var peFile = new PEReader(stream))
                 {
+                    bool hasMetadata = false;
+                    try
+                    {
+                        hasMetadata = peFile.HasMetadata;
```
Comment (BenVillalobos) [inline src/Tasks/AssemblyDependency/AssemblyInformation.cs:367]: nit: carry the comment over that explains why we do this.

**Review by BenVillalobos — COMMENTED** [review]
If we're checking this here, we should add the same check in MetadataReader.cs.
```c#
                    if (_peReader != null)
                    {
                        if (_peReader.HasMetadata)
                        {
                            _reader = _peReader.GetMetadataReader();
                        }
                    }
```

**Comment by FiniteReality** [comment]
> we should add the same check in MetadataReader.cs.

It seems MetadataReader already has a try/catch surrounding the HasMetadata check, which closes the reader, meaning `MetadataReader.Create()` returns `null`.

File: src/Tasks/AssemblyDependency/AssemblyInformation.cs (hunk 1 above)
Comment (FiniteReality) [reply src/Tasks/AssemblyDependency/AssemblyInformation.cs:369]: I've copied over both comments.

**Comment by ladipro** [comment]
This looks good although I wonder if pulling the code into a helper and calling it from both AssemblyNameExtension and AssemblyInformation wouldn't be slightly more elegant.

**Comment by FiniteReality** [comment]
> This looks good although I wonder if pulling the code into a helper and calling it from both AssemblyNameExtension and AssemblyInformation wouldn't be slightly more elegant.

So maybe an extension method `PEReader.ThrowIfMissingMetadata()`?

**Review by cdmihai — CHANGES_REQUESTED** [review]
I think the "prescribed / designed" way of referencing non-CLR projects is [telling the sdk to not treat the ProjectReference output as a managed assembly](https://github.com/dotnet/msbuild/issues/6200#issuecomment-803057767). But I'm not 100% sure, I couldn't find any docs on the expected way to do this (@KathleenDollard @dsplaisted). Might be safer to go down that route than to allow native assemblies to flow into RAR which might have unknown, cascading consequences for the downstream code that reads RAR outputs.

**Comment by ladipro** [comment]
How does RAR behave if you pass a non-PE file?

**Comment by FiniteReality** [comment]
> How does RAR behave if you pass a non-PE file?

Generally, a BadImageFormatException is thrown, which RAR catches and aborts with "File does not contain managed metadata". For example:
```
         Primary reference "HelloWorldLibraryC".
             Could not find dependent files. Assembly file '/home/dev/Development/Finite.Cpp.Sdk/artifacts/bin/samples/HelloWorldLibraryC/Debug/HelloWorldLibraryC.so' could not be opened -- PE image doesn't contain managed metadata.
             Could not find dependent files. Image is too small.
             Resolved file path is "/home/dev/Development/Finite.Cpp.Sdk/artifacts/bin/samples/HelloWorldLibraryC/Debug/HelloWorldLibraryC.so".
             Reference found at search path location "".
             The ImageRuntimeVersion for this reference is "".
```

**Comment by KirillOsenkov** [comment]
Not sure if relevant, but Microsoft.Build.Tasks.Core.dll has an `IsManagedAssembly()` method:
https://source.dot.net/#Microsoft.Build.Tasks.Core/ManifestUtil/PathUtil.cs,ef4ea3875d69614b

Also here's a way to find out if an assembly is managed without throwing:
https://github.com/KirillOsenkov/MetadataTools/blob/main/src/PEFile/IsManagedAssembly.cs

File: src/Tasks/AssemblyDependency/AssemblyInformation.cs [hunk 2]
```diff
@@ -361,6 +361,25 @@ private void CorePopulateMetadata()
...
+                        // the assembly doesn't have metadata.
+                        hasMetadata = peFile.HasMetadata;
+                    }
+                    finally
+                    {
+                        // If the file does not contain PE metadata, throw BadImageFormatException to preserve
+                        // behavior from AssemblyName.GetAssemblyName(). RAR will deal with this correctly.
+                        if (!hasMetadata)
+                        {
+                            throw new BadImageFormatException(string.Format(CultureInfo.CurrentCulture,
```
Comment (KirillOsenkov) [inline src/Tasks/AssemblyDependency/AssemblyInformation.cs:377]: Not sure what is the behavior of throwing an exception from a finally block when we're in the process of throwing a different exception, should we put an empty catch block before it to suppress the other exception?

**Review by KirillOsenkov — APPROVED** [review]

**Comment by KirillOsenkov** [comment]
@cdmihai the project reference stuff is certainly valid, but I'm questioning whether that should be done as a separate PR? This feels nicely self-contained and relatively isolated and it does fix the issue at hand?

File: src/Tasks/AssemblyDependency/AssemblyInformation.cs (hunk 2 above)
Comment (ladipro) [reply src/Tasks/AssemblyDependency/AssemblyInformation.cs:377]: This is fine, it's well-defined to start handling the new exception while skipping the rest of the finally block.

**Review by cdmihai — APPROVED** [review]

**Comment by Forgind** [comment]
Thanks @FiniteReality!

**Pull request merged**